File: Flex/static/rep_data_processor.js
class RepDataProcessor {
  constructor() {
      this.reset();
  }
  
  reset() {
      
      this.reps = [];
      this.currentRep = {
          rep_count: 0,
          max_pitch: -Infinity,
          max_gz_up: -Infinity,  
          max_gz_down: Infinity, 
          max_az: -Infinity
      };
  }
  
  update(pitch, gz, az) {
      
      if (pitch > this.currentRep.max_pitch) {
          this.currentRep.max_pitch = pitch;
      }
      
      
      if (gz > 0) {
          if (gz > this.currentRep.max_gz_up) {
              this.currentRep.max_gz_up = gz;
          }
      } else if (gz < 0) {
          if (gz < this.currentRep.max_gz_down) {
              this.currentRep.max_gz_down = gz;
          }
      }
      
      
      if (az > this.currentRep.max_az) {
          this.currentRep.max_az = az;
      }
  }

  storeRep(repCount) {
      this.currentRep.rep_count = repCount;
      this.reps.push({ ...this.currentRep });  
      this.currentRep = {  
          rep_count: repCount,
          max_pitch: -Infinity,
          max_gz_up: -Infinity,
          max_gz_down: Infinity,
          max_az: -Infinity
      };
  }

  getAggregatedData() {
      return { reps: this.reps };
  }
}

File: raspi/rep_data_processor.py
class RepDataProcessor:
    def __init__(self):
        self.reset()

    def reset(self):
        # Initialize or reset the aggregated metrics for a new rep
        self.min_pitch = float('inf')
        self.max_pitch = -float('inf')
        self.max_gz_up = -float('inf')   # For upward motion (assuming gz positive means upward)
        self.max_gz_down = float('inf')   # For downward motion (assuming gz negative means downward)
        self.max_az = -float('inf')

    def update(self, pitch, gz, az):
        if pitch < self.min_pitch:
            self.min_pitch = pitch
        if pitch > self.max_pitch:
            self.max_pitch = pitch

        if gz > 0:
            if gz > self.max_gz_up:
                self.max_gz_up = gz
        elif gz < 0:
            if gz < self.max_gz_down:
                self.max_gz_down = gz

        # Update max az
        if az > self.max_az:
            self.max_az = az

    def get_aggregated_data(self):
        return {
            "min_pitch": self.min_pitch,
            "max_pitch": self.max_pitch,
            "max_gz_up": self.max_gz_up,
            "max_gz_down": self.max_gz_down,
            "max_az": self.max_az
        }
